ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,864 @@
+  private final Time time;
+  private final UUID sessionId;
+  private final UUID incarnationId;
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  private volatile boolean isActive = false;
+  private PersistentIndex srcIndex;
+
+  private Log tgtLog;
+  private PersistentIndex tgtIndex;
+  private long numSwapsUsed;
+  private StoreFindToken recoveryStartToken = null;
+  private CompactionLog compactionLog;
+  private volatile CountDownLatch runningLatch = new CountDownLatch(0);
+
+  /**
+   * Constructs the compactor component.
+   * @param dataDir the directory where all compactions need to be run.
+   * @param storeId the unique ID of the store.
+   * @param storeKeyFactory the {@link StoreKeyFactory} to generate {@link StoreKey} instances.
+   * @param config the {@link StoreConfig} that defines configuration parameters.
+   * @param metrics the {@link StoreMetrics} to use to record metrics.
+   * @param diskIOScheduler the {@link DiskIOScheduler} to schedule I/O.
+   * @param srcLog the {@link Log} to copy data from.
+   * @param scheduler the {@link ScheduledExecutorService} to use.
+   * @param recovery the {@link MessageStoreRecovery} to use to recover the index and log.
+   * @param time the {@link Time} instance to use.
+   * @param sessionId the sessionID of the store.
+   * @param incarnationId the incarnation ID of the store.
+   */
+  BlobStoreCompactor(String dataDir, String storeId, StoreKeyFactory storeKeyFactory, StoreConfig config,
+      StoreMetrics metrics, DiskIOScheduler diskIOScheduler, Log srcLog, ScheduledExecutorService scheduler,
+      MessageStoreRecovery recovery, Time time, UUID sessionId, UUID incarnationId) {
+    this.dataDir = new File(dataDir);
+    this.storeId = storeId;
+    this.storeKeyFactory = storeKeyFactory;
+    this.config = config;
+    this.metrics = metrics;
+    this.srcLog = srcLog;
+    this.diskIOScheduler = diskIOScheduler;
+    this.scheduler = scheduler;
+    this.recovery = recovery;
+    this.time = time;
+    this.sessionId = sessionId;
+    this.incarnationId = incarnationId;
+  }
+
+  /**
+   * If a compaction was in progress during a crash/shutdown, fixes the state so that the store is loaded correctly
+   * and compaction can resume smoothly on a call to {@link #resumeCompaction()}. Expected to be called before the
+   * {@link PersistentIndex} is instantiated in the {@link BlobStore}.
+   * @throws IOException if the {@link CompactionLog} could not be created or if commit or cleanup failed.
+   * @throws StoreException if the commit failed.
+   */
+  void fixStateIfRequired() throws IOException, StoreException {
+    if (CompactionLog.isCompactionInProgress(dataDir.getAbsolutePath(), storeId)) {
+      compactionLog = new CompactionLog(dataDir.getAbsolutePath(), storeId, storeKeyFactory, time);
+      CompactionLog.Phase phase = compactionLog.getCompactionPhase();
+      switch (phase) {
+        case COMMIT:
+          commit(true);
```

Comment:
For your first question, I found this setup more intuitive. It lists everything that needs work in the beginning and everything that doesn't need any work at the bottom.

Added details about the phases in `resumeCompaction()`